feature/kodb-util clean/import functions WIP

kenner2 opened the pull request:
# kodb-util
`kodb-util` is a database utility tool created to help in the development of the OpenKO project.

The utility aims to:

1. Create/Import the OpenKO database, users, and stored procedures in the OpenKO-db project.
2. TODO: Update the files in the OpenKO-db project via extract functions.

## Program configuration
Configure your database connection information in `config/config.yaml`; As this file is on gitignore, a template is provided in `config/config.yaml.template`

This utility mutates the `knight` user as part of its import and export functionality.  For local development you 
can use your `sa` login; otherwise you'll need to configure a user with similar permissions. 

## Dependencies
This utility is programmed with Go 1.24+.  You'll need to install the language and add it to your PATH. See https://go.dev/doc/install

If Go is correctly installed on your path, you should be able to run `go version` in your terminal and get version
information output:
```
PS C:\> go version
go version go1.24.1 windows/amd64
```

The following commands assume that you have a terminal open in the root folder of the project.

To download project dependencies, run:
```shell
go mod download
```

To run the application, run:
```shell
go run kodb-util.go
```

Without any arguments, you should get a usage prompt like this:
```
|---------------------------|
| OpenKO Database Utilities |
|---------------------------|
No arguments provided:
Usage of kodb-util.exe:
  -clean
        Clean drops the databaseConfig.dbname database and removes the knight user
  -import
        Runs clean and imports OpenKO-db files
```

## Building the utility program
In order to build `kodb-util.exe`, run the following command in this directory:
```shell
go build
```
## TODOs
* Export functionality

Commit:
feature/kodb-util clean/import functions WIP

Commit:
feature/kodb-util Implement transaction fencing for batches and database name specification in the connection string

kenner2 commented:
Did some light refactoring to add DB fencing to the batched SQL and add support for changing DBNames in the connection string.  Data is now loading into correct databases.  I need to do testing to see how the loaded data actually works with the OpenKO server programs, but as it stands:

Example program output:  [exampleOutput.txt](https://github.com/user-attachments/files/20760812/exampleOutput.txt)

OpenKO DB Created with tables/data:
![{70C628AD-899D-490D-9FDC-01D68E4D0CEB}](https://github.com/user-attachments/assets/3303d20c-c426-4e82-8a13-20aa36a389e4)

Views:
![{892D9DAE-3010-40EC-8CA7-5DB00C3DC405}](https://github.com/user-attachments/assets/515e0616-c4b8-4a78-bdba-2d4fcfbeb11f)

Stored Procs:
![{67D4D0FA-66AF-4A9D-91B0-6D5A8AF333CA}](https://github.com/user-attachments/assets/b89a0f4b-3f59-4e55-8b23-a97a72873f0a)

Schemas:  knight schema is created under OpenKO db.  I'm not 100% clear if this should be in OpenKO or Master, or what it's purpose is other than being the default schema for the knight user.  I don't have this on my 1298 database backups.

User/Login:  knight is created under Master DB.  Not sure if functional for server apps / odbc yet.

kenner2 commented on Databases/OpenKO.sql:
```diff
@@ -0,0 +1,109 @@
+USE [master]
+GO
+
+CREATE DATABASE [OpenKO]
```
Long run this would best be done as a template file, where OpenKO is replaced by %s for config substitution.  Would need to identify where [OpenKO] is referenced in existing dump files and update those to be templated as well, at that time.

kenner2 commented on Tables/HOME.sql:
Since the diff isn't loading:  This file contained unbalanced parenthesis:
![{7E399820-ED4D-43DA-8DE0-F74F79565E65}](https://github.com/user-attachments/assets/4faed889-7337-4ca8-984f-c9ceb6bf953e)

twostars replied:
This is because it generates with full columns and I manually stripped them out for the sake of reducing massive dumps like item.sql. It seems I broke this case.

Ideally we should be automatically generating this via tool, since I couldn't really do it via MSSQL in a nice enough way.

kenner2 commented on Tables/KNIGHTS_RATING.sql:
File was using IDENTITY INSERT ON but not specifying column names in insert, which threw mssql errors:
![{40C10CB8-4206-457D-812A-6C0DB4F196DB}](https://github.com/user-attachments/assets/f2dd2ea0-541e-4bd1-9aaa-0fbcebf013ac)

kenner2 commented on Tables/SERVERS.sql:
Updated to correct IDENTITY INSERT ON syntax
![{7100365C-468F-4E06-AE51-C113DB5DB8D9}](https://github.com/user-attachments/assets/4177058f-47dd-451d-bd1f-4470439a569c)

kenner2 commented on Tables/USER_SAVED_MAGIC.sql:
Corrected unbalanced parenthesis:
![{992320B1-3FE5-4665-AF45-5EB1B5B1F6A1}](https://github.com/user-attachments/assets/4fcc340c-c439-4b14-b53a-90296aee5bc1)

Commit:
feature/kodb-util comment cleanup

kenner2 commented:
Circling back around to confirm that the knight user works and all OpenKO server programs launch successfully with data imported from the utility. Was able to create a character, log into the game, buy/equip/store items, kill stuff, level up, etc.

Only thing I had to change was the default IP in SERVERS.  We should probably change that to 127.0.0.1 as a default

twostars commented:
The SQL errors are a result of the export and manual touch up to reduce filesize. This is why this really needs to be done automatically via tool. Ugh.

kenner2 commented:
> The SQL errors are a result of the export and manual touch up to reduce filesize. This is why this really needs to be done automatically via tool. Ugh.

🤷  Gotta start somewhere.   It only took a few hours to knock out coding a basic import script.  These dumps you generated wouldn't be too hard to convert to template files, and then have an export function fill the data in.   I think it's a pretty decent starting point

kenner2 commented:
Just as a little cross-talk from the other thread:

> The problem however is with the database everyone uses for this. It's a singular database with only the game implementation.
>
>Without a way of keeping their database up-to-date with relevant changes, this is somewhat problematic. It might have to be tweaked to just manually lookup the data directly and handle it on the login server's side until we have database stuff sorted and can make appropriate changes there...

It would be pretty easy to refactor this to support a multi-db implementation.  Would just need to do some folder restructuring so we'd have:
-Databases
-- ACCOUNTS
--- Tables
--- Views
--- Stored Procedures
-- GAME
--- Tables
--- etc...

Commit:
feature/kodb-util Updated file encodings of SQL batch files from UTF-16LE to UTF-8.  Removed UTF-16LE conversion code and dependency library.  Updated SERVERS.sql entry to use localhost.  Updated WEBPAGE_ADDRESS to point to ddg.gg (duckduckgo) to avoid default website trackers.  Added personal login to TB_USERS.sql

kenner2 commented:
It's in the commit message, but to make sure it's seen:

- Converted all the SQL batch files to UTF-8, reducing their file size by ~50%
- Updated the SERVERS.sql entry to default the name to "OpenKO" and host to "127.0.0.1"
- Updated WEBPAGE_ADDRESS.sql to default to "ddg.gg" (duckduckgo search) so that no third party trackers are being activated in the case that someone's client opens the default link.  Would have just put the github link in but the max URL size is 10 characters on that column.
- Added a login for myself to TB_USER.sql

Did a clean/import on my local instance and did a basic test run without issues.   If you want me to revert any of the SQL changes above, just let me know.

twostars commented:
That's fine. WEBPAGE_ADDRESS isn't even used, as far as the URL itself goes. That's all exclusively client-sided.

The server will tell it which ID to use, but the URL it loads will be from the client TBL not the server. I'm not sure why it includes a URL in the DB. Even the column's size is wrong for 1.298 here (this is just people recreating tables without looking too deep into it). It binds to a max of 255 characters (although it *stores* only 100, so theoretically 99 characters long + null-terminator).

twostars commented:
Unfortunately this repo's screwed as far as those files go now since they were UTF-16 at some point (I believe .gitattributes won't necessarily behave correctly since it was once UTF-16 at some point) but I didn't really intend on this being a thing anyway, so we can just recreate this stuff again later in a proper repo.

I don't know if you're looking for me to merge this stuff or not. I personally don't really care too much either way, since it was just an experimental repo to see what I'd end up doing with this stuff for the sake of the real one, but yeah.

kenner2 commented:
Are you going to use the files from this repo as a baseline for the new repo?  I would imagine the .gitattributes would recalculate and we'd get actual diffs on the sql batches in the new repo.

I'm good to merge either way.  I want to add outer Tx fencing but that's tricky with db swapping.  For the export function impl I'm going to refactor to use GORM (https://github.com/go-gorm/gorm) for the database driver, since that gives struct binding support.  That should also give me the ability to do nested Tx fences.  I used GORM in a previous job - it's a solid library.  I can repurpose ko-codegen to spit out an object library with GORM bindings for the export job.  A bit of lifting but it should be a really solid solution once done.  Either way - I want to do the export stuff on a new PR/Branch just so I have a clean set of diffs to look at while self-reviewing.

I'm guessing the refactor to the OpenKO codebase to revert to the original 1068 code is going to take a bit of time, and splitting into multiple DBs is going to be part of that process.  If you want to open a repo under the OpenKO organization with this single-db impl that works for the current server structure, we can:

- Open a branch under the new repo for multi-database support that becomes the main branch when the 1068 refactor gets merged to OpenKO's master branch
- I'll work on export functionality for single-db while the 1068 refactor is in progress.  As the 1068 multi-db becomes available I'll refactor the kodb-util program to support it
- Cut an OpenKO-db branch prior to merging multi-db to main that preserves the single-db implementation for anyone who is using that setup, so that both approaches have something out there in terms of SCM databases/data sets.

twostars commented:
Yeah, when they're introduced into a fresh repo they'll behave fine. We can use them, but I anticipated just using more consistently exported dumps rather than the MSSQL exported + manually touched up stuff.

So I'll probably hold off on creating a repo until we have those baseline exported; then we can use those from the get-go, rather than reworking all the dumps.

I feel like ultimately, the tool should probably go in its own repo perhaps, and have the binary accessible for general use in some fashion, but I'm not sure how I envision that working. Perhaps another repository for the binaries, submoduled into the KnightOnline repo alongside this repo and hook things up to just run the binary for installation or updates (for general use). That still somewhat implies manual usage, but if we had a versioning table that the server fetched, we could pair associated versions with our server builds and insist they update so the database schema matches.

kenner2 commented:
I figured we might want to relocate it. It'll be easy to add a config param to the tool that specifies the base directory of the OpenKO-db project.   I just opened a directory and started coding to see how well this would even work - it all went smoother than expected.  

Sure - we can use git tags (for devs that want to compile/run locally) and github releases (for folks that want to download a precompiled binary) and sync verison numbers between the projects.  

For a github release I think it would make sense to have a combined release of the SQL stuff + the tool, where the tool is preconfigured to run in the base directory of the release.  Fewer steps for a user of a release download are fewer false issues reported in the repo 🙂 

If you want to take the stuff in kodb-util and use it as the baseline for the tool repo in the organization projects, go for it.  I assume you'd want it under the organization and not my personal account; I'd prefer that as well.   I'm not sure how organization permissions work, but if you can give me merge permissions for the tool repo when you make it, that'd be great.

twostars commented:
I'm not too sure about having them manually download it and run it like that. Like that's fine, but I was thinking more in terms of centralising it under the main repo's banner, and having a submodule point to a repo with the relevant binaries. Probably a script (or symlink?) in the root of the main repo to run it.

That way it's just "there" and "just works" (in theory - I don't know much about Go binaries and their dependencies, though).

So they clone the repo, and then as per the readme and/or the server, it can prompt them to just run it to get their database setup.

As for kodb-util, I think I can probably add you as a contributor or maintainer or whatever of a given repo.

Edit:
I'll merge this in and do that one but we can continue to discuss here anyway.

twostars commented:
Added you over at https://github.com/Open-KO/kodb-util

kenner2 commented:
I've never used sub-modules for a project before; I'll read up on them in a bit so I'm on the same page.  High-level I understand where you're going, though.  No concerns with it.

Go programs build into self-contained exe files - the only thing they would be dependent on are non-code resources (like pretty much any other exe).  So yeah - it should just work; would just need access to it's config file. 

If you're a little iffy about using Go for some of the tooling, I get it.  I was worried about it being some flavor-of-the-month hipster language when I was preparing to interview for a prior job that used it.  

I went through the processing of benchmarking it against C++ and Java in solving the nQueens problem, using their respective concurrency libraries.  C++ was best for memory usage hands-down, but was also the slowest to execute (I have thoughts on this - it's likely due to language design avoiding using heap memory in Java/Go.  Using stack memory is so much faster on modern CPUs).  Java and Go had similiar execution times until you increased thread count past a certain point, at which point Go blew Java out of the water (newer versions of the JVM finally have virtual threads without so much overhead, so this may have leveled out in JDK 21+).  Java was the worst on memory, Go was in the middle.   Anyway - performance wise it's pretty good.  A bit more memory usage, but you don't have to manually manage pointers / deconstructors.

It was designed by some real CS powerhouses, too:  [Robert Griesemer](https://en.wikipedia.org/wiki/Robert_Griesemer), [Rob Pike](https://en.wikipedia.org/wiki/Rob_Pike), and [Ken Thompson](https://en.wikipedia.org/wiki/Ken_Thompson).  I watched a few talks from these guys and a lot of their inspiration in writing the language was to make it into a successor of C; learning from all the issues that they had with it along the way, as well as addressing common pitfalls C developers commonly fall into.  (Example: a switch's case statement is assumed to end with "break" by default rather than fallthrough.  If you want a case to fallthrough into the next, you have to invoke a fallthrough keyword)

The only thing I ever really worried about with Go is that it's technically maintained/funded by Google, and they love to send stuff to their graveyard.  I think Go has become widely adopted enough at this point that it shouldn't be much of a concern, though. 

Anyway - a lot of words you may have no interest in.  Go is a lot like C with some slight syntax differences.  Dependency management is very easy with it.

twostars commented:
As a language, I don't mind it. When I talk about its dependencies, I'm referring to like, any runtimes required to be installed for it or whatever. But you said they're self-contained so that *should* be OK. I'd have to open up a binary to have a quick look to be certain, but yeah, the intention is that it should "just work" for people so we don't have to deal with some people not being able to run it, etc.

Requiring others to install Go and build it is a bit of a harder ask of course, and that was my original concern. But yeah, if the executables "just work" we're good.

kenner2 commented:
>  I'd have to open up a binary to have a quick look to be certain, but yeah, the intention is that it should "just work" for people so we don't have to deal with some people not being able to run it, etc.

If you're looking for imports, the only one that's showing in the disassembler is kernel32.dll

twostars commented:
That's perfect :)

kenner2 commented:
> I'm not too sure about having them manually download it and run it like that. Like that's fine, but I was thinking more in terms of centralising it under the main repo's banner, and having a submodule point to a repo with the relevant binaries. Probably a script (or symlink?) in the root of the main repo to run it.

Read a bit on submodules: 

- The submodule repo would have to be a stand-alone repo for released binaries.  I think this is what you meant, but I want to make sure we're on the same page. If the idea was to make kodb-util a submodule then we would either have to have a binary committed to the repo (I have exe on gitignore) or the mentioned script would need to build the project into an exe (which would require installing Go)
- The user is either going to need to make changes to the config/config.yaml file, or the script is going to need to take user input and update it for them.  The thing that sticks out in the config is username/password; The config assumes sa as the default user.  I would hope everyone is using different/secure passwords.  Could also consider a common config file that the shell scripts read and pass the user/pass from that - I would just need to add CLI parsing for those arguments. 
- kodb-util.exe will take different params for different actions (-clean, -import, -export), so having different shell scripts with the parameters already specified (importDb.bat, exportDb.bat) would probably be best.  
- Should be a non-issue but to point it out ahead of time: if these shell scripts call, say, ./openko-tools/kodb-util.exe, then the folder that the bat files are executed from should be considered the working directory of the running application.  Configuration of the OpenKO-db path will need to take that into consideration.

From my own comments here:
> The user is either going to need to make changes to the config/config.yaml file

I'll refactor the location of this config to be in the same directory as the binary and make the file name specific to the application to prevent collisions with other tools (kodb-utils-config.yaml)

twostars commented:
> The submodule repo would have to be a stand-alone repo for released binaries. I think this is what you meant, but I want to make sure we're on the same page. If the idea was to make kodb-util a submodule then we would either have to have a binary committed to the repo (I have exe on gitignore) or the mentioned script would need to build the project into an exe (which would require installing Go)

This is correct. 

> The user is either going to need to make changes to the config/config.yaml file, or the script is going to need to take user input and update it for them. The thing that sticks out in the config is username/password; The config assumes sa as the default user. I would hope everyone is using different/secure passwords. Could also consider a common config file that the shell scripts read and pass the user/pass from that - I would just need to add CLI parsing for those arguments.

There's actually another option but it should still handle fallback regardless, in case it's *not* set; Windows authentication is supported (mixed mode is enabled by default on install unless you explicitly turn it off). In this case, their local user should automatically just be able to auth without issue.

> Should be a non-issue but to point it out ahead of time: if these shell scripts call, say, ./openko-tools/kodb-util.exe, then the folder that the bat files are executed from should be considered the working directory of the running application. Configuration of the OpenKO-db path will need to take that into consideration.

Could also take away any assumptions and pass a command-line arg for the config path.

kenner2 commented:
> In this case, their local user should automatically just be able to auth without issue.

Yup, just implemented support for that locally.  I was expecting it to prompt for my password or something...but nope, just went through.  Horrifying.  Microsoft, bay-bee.

kenner2 commented:
Added changes discussed in this thread in this PR:  https://github.com/Open-KO/kodb-util/pull/1/files

Time to let this thread go off to pasture

kenner2 commented:
Tahsin added me to one of your servers the other day - added you from there
